3.5/troubleshooting-reduce-memory-footprint.md

```diff
@@ -0,0 +1,353 @@
+physical memory is available. The kernel will only grant up to `vm.max_map_count`
+memory mappings to each process, which defaults to 65530 on many Linux
+environments.
+
+Another issue when running jemalloc with `vm.overcommit_memory` set to **2** is
+that for some workloads the amount of memory that the Linux kernel tracks as
+“*committed memory*” also grows over time and never decreases. Eventually,
+*arangod* may not get any more memory simply because it reaches the configured
+overcommit limit (physical RAM * `overcommit_ratio` + swap space).
+
+The solution is to modify the value of `vm.overcommit_memory` from **2** to
+either **0** or **1**. This will fix both of these problems. We still observe
+ever-increasing *virtual memory* consumption when using jemalloc regardless of
+the overcommit setting, but in practice this should not cause any issues.
+Adjusting the value to either **0** or **1** should improve the situation. **0**
+is the Linux kernel default and also the setting we recommend.
+
+For the sake of completeness, let us also mention another way to address the
+problem: use a different memory allocator. This requires to compile ArangoDB
+from the source code without jemalloc (`-DUSE_JEMALLOC=Off` in the call to cmake).
+With the system’s libc allocator you should see quite stable memory usage. We
+also tried another allocator, precisely the one from `libmusl`, and this also
+shows quite stable memory usage over time. What holds us back to change the
+bundled allocator are that it is a non-trivial change and because jemalloc has
+very nice performance characteristics for massively multi-threaded processes
+otherwise.
+
+Testing the Effects of Reduced I/O Buffers
+------------------------------------------
+
+![Performance Graph](images/performance_graph.png)
+
+- 15:50 – Start bigger import
+- 16:00 – Start writing documents of ~60 KB size one at a time
+- 16:45 – Add similar second writer
+- 16:55 – Restart ArangoDB with the RocksDB write buffer configuration suggested above
+- 17:20 – Buffers are full, write performance drops
+- 17:38 – WAL rotation
+
+What you see in above performance graph are the consequences of restricting the
+write buffers. Until we reach a 90% fill rate of the write buffers the server
+can almost follow the load pattern for a while at the cost of constantly
+increasing buffers. Once RocksDB reaches 90% buffer fill rate, it will
+significantly throttle the load to ~50%. This is expected according to the
+[upstream documentation](https://github.com/facebook/rocksdb/wiki/Write-Buffer-Manager):
+
+> […] a flush will be triggered […] if total mutable memtable size exceeds 90%
+> of the limit. If the actual memory is over the limit, more aggressive flush
+> may also be triggered even if total mutable memtable size is below 90%.
+
+Since we only measured the disk I/O bytes, we don’t see that the document save
+operations also doubled in request time.
+
+Summary
```

Comment:
again blog post or tutorial style

Reply:
moved to the top to act as an introduction